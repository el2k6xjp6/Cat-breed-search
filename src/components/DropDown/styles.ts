import styled from "styled-components";

export const DropDownContainer = styled.div`
  cursor: pointer;
  width: 200px;
  height: 40px;
`;

export const DropDownHeader = styled.div`
  position: relative;
  border-radius: 4px;
  border: solid 1px #c4c4c4;
  height: 40px;
  display: flex;
  align-items: center;
  background: white;
  :hover {
    border: solid 1px black;
  }
`;

const Text = styled.div`
  text-indent: 15px;
`;

export const Placeholder = styled(Text)`
  color: #aaaaaa;
`;

export const HeaderText = styled(Text)`
  color: #030303;
`;

export const Arrow = styled.div<{ isOpen: boolean }>`
  position: absolute;
  right: 10px;
  color: #909090;
  ${({ isOpen }) =>
    isOpen
      ? "transform: rotate(180deg); transition: transform .2s ease-in-out"
      : "transform: rotate(360deg); transition: transform .2s ease-in-out"};
`;

export const DropDownListContainer = styled.div`
  position: absolute;
  z-index: 100;
  margin-top: 10px;
`;

export const DropDownList = styled.ul`
  box-shadow: rgb(0 0 0 / 8%) 0px 4px 12px;
  border-radius: 8px;
  margin: 0;
  min-width: fit-content;
  min-height: fit-content;
  max-height: 250px;
  background: white;
  overflow-y: scroll;
  box-sizing: border-box;
  list-style-type: none;
  text-align: left;
  padding: 10px 5px 10px 10px;
  ::-webkit-scrollbar {
    width: 5px;
    background: rgba(0, 0, 0, 0);
  }
  ::-webkit-scrollbar-track {
    background: rgba(0, 0, 0, 0);
    width: 5px;
    position: relative;
  }
  ::-webkit-scrollbar-thumb:hover {
    background: rgba(0, 0, 0, 0);
  }
`;

export const ListItem = styled.li`
  display: flex;
  align-items: center;
  border-radius: 4px;
  font-size: 16px;
  height: 36px;
  line-height: 36px;
  list-style-type: none;
  padding: 0 12px;
  cursor: pointer;
  :hover {
    background: #dddddd;
  }
`;

export const ListItemText = styled.div`
  @media (hover: hover) and (pointer: fine) {
    ${ListItem}: hover & {
      color: black;
    }
  }
`;
