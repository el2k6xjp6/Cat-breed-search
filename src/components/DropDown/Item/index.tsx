import React from "react";
import { ListItem, ListItemText } from "../styles";

export type optionType = {
  text: string;
  value: number;
};

type DropDownListProps = {
  option: optionType;
  index: number;
  onOptionClicked: (option: optionType) => void;
};

const Item = ({ option, index, onOptionClicked }: DropDownListProps) => {
  const { text, value } = option;
  return (
    <ListItem onClick={() => onOptionClicked(option)} key={index} value={value}>
      <ListItemText>{text}</ListItemText>
    </ListItem>
  );
};

export default Item;
