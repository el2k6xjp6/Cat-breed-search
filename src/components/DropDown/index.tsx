import React, { useRef, useEffect, useCallback, useMemo } from "react";
import {
  DropDownContainer,
  DropDownHeader,
  HeaderText,
  DropDownListContainer,
  DropDownList,
  Arrow,
  Placeholder,
} from "./styles";
import Item, { optionType } from "./Item";

interface Props {
  selectedText: string;
  options: Array<optionType>;
  placeHolder: string;
  handleOptionChange: (value: number) => void;
  validateSpy?: () => void;
}

function DropDown(props: Props) {
  const dropdownRef = useRef() as React.RefObject<HTMLDivElement>;
  const [toggle, setToggle] = React.useState(false);

  const handleClickOutside = useCallback(
    (event: MouseEvent) => {
      if (
        dropdownRef.current &&
        dropdownRef.current != null &&
        !dropdownRef.current.contains(event.target as Node)
      ) {
        setToggle(false);
      }
    },
    [dropdownRef]
  );

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [handleClickOutside]);

  const toggling = (): void => setToggle(!toggle);

  const onOptionClicked = useMemo(
    () => (option: optionType) => {
      if (props.validateSpy != null) {
        props.validateSpy();
      }
      const { value } = option;
      setToggle(false);
      props.handleOptionChange(value);
    },
    []
  );

  return (
    <DropDownContainer ref={dropdownRef}>
      <DropDownHeader id="DropDownHeader" onClick={toggling}>
        {props.selectedText == null ? (
          <Placeholder>{props.placeHolder}</Placeholder>
        ) : (
          <HeaderText>{props.selectedText}</HeaderText>
        )}
        <Arrow id="DropDownArrow" isOpen={toggle}>
          â–¼
        </Arrow>
      </DropDownHeader>
      {toggle && (
        <DropDownListContainer>
          <DropDownList>
            {props.options.map((option: optionType, index: number) => (
              <Item
                key={"item-" + index}
                option={option}
                index={index}
                onOptionClicked={onOptionClicked}
              />
            ))}
          </DropDownList>
        </DropDownListContainer>
      )}
    </DropDownContainer>
  );
}

export default React.memo(DropDown);
