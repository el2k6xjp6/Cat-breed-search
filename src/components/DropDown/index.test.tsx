import React from "react";
import { shallow, mount } from "enzyme";
import DropDown from ".";
import "jest-styled-components";
import renderer from "react-test-renderer";

describe("test DropDown", () => {
  const mockProps = {
    selectedText: "",
    options: [{ text: "test", value: 0 }],
    placeHolder: "",
    handleOptionChange: jest.fn(),
  };

  test("Snapshot for closing DropDown", () => {
    const tree = renderer.create(<DropDown {...mockProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Snapshot for opening DropDown", () => {
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [true, setState]);
    const tree = renderer.create(<DropDown {...mockProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("toggle open", () => {
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [false, setState]);
    const wrapper = shallow(<DropDown {...mockProps} />);
    wrapper.find("[id='DropDownHeader']").simulate("click");
    expect(setState).toHaveBeenCalledWith(true);
  });

  test("toggle close", () => {
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [true, setState]);
    const wrapper = shallow(<DropDown {...mockProps} />);
    wrapper.find("[id='DropDownHeader']").simulate("click");
    expect(setState).toHaveBeenCalledWith(false);
  });

  test("click option", () => {
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [true, setState]);
    const validateSpy = jest.fn();
    const wrapper = mount(
      <DropDown {...mockProps} validateSpy={validateSpy} />
    );
    wrapper.find("[id='option 0']").first().simulate("click");
    expect(validateSpy).toHaveBeenCalledTimes(1);
  });
});

export {};
