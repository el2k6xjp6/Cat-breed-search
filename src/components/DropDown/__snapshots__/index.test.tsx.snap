// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test DropDown Snapshot for DropDown close 1`] = `
.c0 {
  cursor: pointer;
  width: 200px;
  height: 40px;
}

.c1 {
  position: relative;
  border-radius: 4px;
  border: solid 1px #c4c4c4;
  height: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: white;
}

.c1:hover {
  border: solid 1px black;
}

.c2 {
  text-indent: 15px;
}

.c3 {
  color: #030303;
}

.c4 {
  position: absolute;
  right: 10px;
  color: #909090;
  -webkit-transform: rotate(360deg);
  -ms-transform: rotate(360deg);
  transform: rotate(360deg);
  -webkit-transition: -webkit-transform .2s ease-in-out;
  -webkit-transition: transform .2s ease-in-out;
  transition: transform .2s ease-in-out;
}

<div
  className="c0"
>
  <div
    className="c1"
    id="DropDownHeader"
    onClick={[Function]}
  >
    <div
      className="c2 c3"
    >
      
    </div>
    <div
      className="c4"
      id="DropDownArrow"
    >
      ▼
    </div>
  </div>
</div>
`;

exports[`test DropDown Snapshot for DropDown open 1`] = `
.c0 {
  cursor: pointer;
  width: 200px;
  height: 40px;
}

.c1 {
  position: relative;
  border-radius: 4px;
  border: solid 1px #c4c4c4;
  height: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: white;
}

.c1:hover {
  border: solid 1px black;
}

.c2 {
  text-indent: 15px;
}

.c3 {
  color: #030303;
}

.c4 {
  position: absolute;
  right: 10px;
  color: #909090;
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
  -webkit-transition: -webkit-transform .2s ease-in-out;
  -webkit-transition: transform .2s ease-in-out;
  transition: transform .2s ease-in-out;
}

.c5 {
  position: absolute;
  z-index: 100;
  margin-top: 10px;
}

.c6 {
  box-shadow: rgb(0 0 0 / 8%) 0px 4px 12px;
  border-radius: 8px;
  margin: 0;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
  min-height: -webkit-fit-content;
  min-height: -moz-fit-content;
  min-height: fit-content;
  max-height: 250px;
  background: white;
  overflow-y: scroll;
  box-sizing: border-box;
  list-style-type: none;
  text-align: left;
  padding: 10px 5px 10px 10px;
}

.c6::-webkit-scrollbar {
  width: 5px;
  background: rgba(0,0,0,0);
}

.c6::-webkit-scrollbar-track {
  background: rgba(0,0,0,0);
  width: 5px;
  position: relative;
}

.c6::-webkit-scrollbar-thumb:hover {
  background: rgba(0,0,0,0);
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  font-size: 16px;
  height: 36px;
  line-height: 36px;
  list-style-type: none;
  padding: 0 12px;
  cursor: pointer;
}

.c8:hover {
  background: #dddddd;
}

@media (hover:hover) and (pointer:fine) {
  .c7:hover .c9 {
    color: black;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
    id="DropDownHeader"
    onClick={[Function]}
  >
    <div
      className="c2 c3"
    >
      
    </div>
    <div
      className="c4"
      id="DropDownArrow"
    >
      ▼
    </div>
  </div>
  <div
    className="c5"
  >
    <ul
      className="c6"
    >
      <li
        className="c7 c8"
        id="option 0"
        onClick={[Function]}
        value={0}
      >
        <div
          className="c9 "
        >
          test
        </div>
      </li>
    </ul>
  </div>
</div>
`;
